### ISSUE TITLE
Same query results are inconsistent on TiKV and TiFlash when involving `CAST YEAR`.

### ISSUE THREAD

**Issue by fenfeng9** [pingcap/tidb#57850]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
prepare data
```sql
drop table if exists t0;
CREATE TABLE t0(c0 int );
INSERT INTO t0(c0) VALUES (1), (2);
ALTER TABLE t0 SET TIFLASH REPLICA 1;
```
execute query
```sql
SELECT /*+ read_from_storage(tikv[t0]) */  t0.c0 FROM t0 WHERE (1>(((t0.c0)>>(CAST(1.469047418E9 AS YEAR)))));
SELECT /*+ read_from_storage(tiflash[t0]) */  t0.c0 FROM t0 WHERE (1>(((t0.c0)>>(CAST(1.469047418E9 AS YEAR)))));
```

### 2. What did you expect to see? (Required)
The results of the two queries are equal.

### 3. What did you see instead (Required)
Query1: 
![Image](https://github.com/user-attachments/assets/c031cedc-8ee0-4d7f-a287-c888f376e641)
Query2: empty
### 4. What is your TiDB version? (Required)

Release Version: v8.4.0
Edition: Community
Git Commit Hash: 1a9f0fa25580907d3266dc3edf000ba24379a600
Git Branch: HEAD
UTC Build Time: 2024-11-07 15:18:43
GoVersion: go1.23.2
Race Enabled: false
Check Table Before Drop: false
Store: tikv

**Comment by yibin87** [comment]
BTW, are you doing some research on databases?

**Comment by fenfeng9** [comment]
Yes, I am doing some simple tests. Thank you for taking the time to confirm these bugs. @yibin87
